servers/quarkus-common/src/main/java/org/projectnessie/quarkus/providers/storage/CacheInvalidationReceiver.java

```diff
@@ -0,0 +1,172 @@
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        },
+        senderId,
+        token);
+  }
+
+  void cacheInvalidations(
+      RoutingContext rc,
+      Supplier<CacheInvalidations> invalidations,
+      String senderId,
+      String token) {
+    if (token == null || !validTokens.contains(token)) {
+      LOGGER.warn("Received cache invalidation with invalid token {}", token);
+      responseInvalidToken(rc);
+      return;
+    }
+    if (serverInstanceId.equals(senderId)) {
+      LOGGER.trace("Ignoring invalidations from local instance");
+      responseNoContent(rc);
+      return;
+    }
+    if (!"application/json".equals(rc.request().getHeader("Content-Type"))) {
+      LOGGER.warn("Received cache invalidation with invalid HTTP content type");
+      responseInvalidContentType(rc);
+      return;
+    }
+
+    List<CacheInvalidations.CacheInvalidation> invs;
+    try {
+      invs = invalidations.get().invalidations();
+    } catch (RuntimeException e) {
+      responseServerError(rc);
+      return;
+    }
+
+    for (CacheInvalidations.CacheInvalidation invalidation : invs) {
+      switch (invalidation.type()) {
+        case CacheInvalidations.CacheInvalidationEvictObj.TYPE:
+          CacheInvalidations.CacheInvalidationEvictObj putObj =
+              (CacheInvalidations.CacheInvalidationEvictObj) invalidation;
+          invalidator.invalidation.evictObj(putObj.repoId(), objIdFromByteArray(putObj.id()));
+          break;
+        case CacheInvalidations.CacheInvalidationEvictReference.TYPE:
+          CacheInvalidationEvictReference putReference =
+              (CacheInvalidations.CacheInvalidationEvictReference) invalidation;
+          invalidator.invalidation.evictReference(putReference.repoId(), putReference.refName());
+          break;
+        default:
+          // nothing we can do about a new invalidation type here
+          break;
+      }
+    }
+
+    responseNoContent(rc);
+  }
+
+  private void responseServerError(RoutingContext rc) {
+    rc.response().setStatusCode(500).setStatusMessage("Server error parsing request body").end();
+  }
+
+  private void responseInvalidToken(RoutingContext rc) {
+    rc.response().setStatusCode(400).setStatusMessage("Invalid token").end();
+  }
+
+  private void responseInvalidContentType(RoutingContext rc) {
+    rc.response().setStatusCode(415).setStatusMessage("Unsupported media type").end();
+  }
+
+  private void responseNoContent(RoutingContext rc) {
+    rc.response().setStatusCode(204).setStatusMessage("No content").end();
+  }
+
+  @Singleton
+  public static class DistributedCacheInvalidationHolder
+      implements DistributedCacheInvalidationConsumer {
+    DistributedCacheInvalidation invalidation;
+
+    @Override
+    public void accept(DistributedCacheInvalidation distributedCacheInvalidation) {
+      this.invalidation = distributedCacheInvalidation;
```

Comment:
I wonder if it might be simpler to attach the cache (direct java call) to a fiend in `CacheInvalidationReceiver` when the cache is constructed. It could save an indirection, I guess.

Reply:
Simplified in the last commit